File: lib/src/main/java/com/asyncanimator/launcher/pending/AnimationSuccessListener.java
package com.asyncanimator.launcher.pending;

import android.animation.Animator;

/**
 * AnimationSuccessListener — 区分 cancel 和 success 的 listener 基类。
 *
 * <p>对应分析文档 §6.3.4。核心设计：
 * <ul>
 *   <li>cancel 路径把 {@code mCancelled = true}</li>
 *   <li>end 路径上若 mCancelled=true 则不触发 onAnimationSuccess</li>
 *   <li>业务只 override {@link #onAnimationSuccess(Animator)}</li>
 * </ul>
 */
public abstract class AnimationSuccessListener extends ActualEndAnimListener {

    @Override
    public void onAnimationCancel(Animator animator) {
        super.onAnimationCancel(animator);
        mCancelled = true;
    }

    @Override
    public void onAnimationEnd(Animator animator) {
        if (mCancelled) return;
        onAnimationSuccess(animator);
    }

    public abstract void onAnimationSuccess(Animator animator);
}

File: lib/src/main/java/com/asyncanimator/launcher/pending/AnimationSuccessListener.kt
package com.asyncanimator.launcher.pending

import android.animation.Animator

/**
 * AnimationSuccessListener — 区分 cancel 和 success 的 listener 基类。
 *
 * 对应分析文档 §6.3.4。核心设计：
 *
 *  - cancel 路径把 `cancelled = true`
 *  - end 路径上若 cancelled=true 则不触发 onAnimationSuccess
 *  - 业务只 override [onAnimationSuccess]
 */
abstract class AnimationSuccessListener : ActualEndAnimListener() {

    override fun onAnimationCancel(animator: Animator) {
        super.onAnimationCancel(animator)
        cancelled = true
    }

    override fun onAnimationEnd(animator: Animator) {
        if (cancelled) return
        onAnimationSuccess(animator)
    }

    abstract fun onAnimationSuccess(animator: Animator)
}
